File: reef/harness/adapters/pi/requests.ts
// Harness requests: the /reef-harness and /reef-versions commands for reef-pi. The
// person asks in plain words; the request goes to reef with this session's id
// and the installed release through native manual training; the service
// proposer writes the change without requiring inference receipts or a
// feedback report. /reef-versions lists the release chain with each step's
// verdict and request, prints a step's page and, for a pending release, the
// promote action and a trial install, and runs the promote after a
// confirmation. Nothing here writes a mutation. Kept free of annotations on
// purpose: plain JavaScript in a .ts file, so plain node can parse it in CI
// and pi's TS loader accepts it unchanged. Gate episodes set PI_OFFLINE and
// this extension then registers nothing, so the gate never sees the commands.
import { readFileSync } from "node:fs";
import { join } from "node:path";

// The release file the install script and harness_pull write at the tree root.
const RELEASE_FILE = ".reef-harness-release";

function readJson(path) {
  try {
    return JSON.parse(readFileSync(path, "utf8"));
  } catch {
    return null;
  }
}

function message(error) {
  return error instanceof Error ? error.message : String(error);
}

export default function requests(pi) {
  if (process.env.PI_OFFLINE) return; // hermetic episodes never see the command
  const agentDir = process.env.PI_CODING_AGENT_DIR;
  const serviceUrl = process.env.REEF_SERVICE_URL;
  const scenario = process.env.REEF_SCENARIO;
  if (!agentDir || !serviceUrl || !scenario) return;
  // The wrapper relocates the agent into a temp copy and exports the true
  // install root; a tree run directly falls back to the release file beside it.
  const destDir = process.env.REEF_HARNESS_DEST || join(agentDir, "..");

  const reefHeaders = () => {
    const token = process.env.REEF_TOKEN;
    return { "x-reef-scenario": scenario, ...(token ? { authorization: `Bearer ${token}` } : {}) };
  };

  const installedRelease = () => {
    const releaseInfo = readJson(join(destDir, RELEASE_FILE));
    return releaseInfo && typeof releaseInfo.release_id === "string" && releaseInfo.release_id ? releaseInfo.release_id : null;
  };

  pi.registerCommand("reef-harness", {
    description: "Ask reef to grow this harness: /reef-harness <what it should do>",
    handler: async (args, ctx) => {
      const text = (args || "").trim();
      if (!text) {
        ctx.ui.notify("Usage: /reef-harness <what the harness should do>", "warning");
        return;
      }
      const releaseId = installedRelease();
      if (!releaseId) {
        ctx.ui.notify(
          `no ${RELEASE_FILE} release file at ${destDir}: this tree did not come through reef's install channel, ` +
            "so a request cannot name the release it runs; nothing was sent",
          "error",
        );
        return;
      }
      const body = { text, session: ctx.sessionManager.getSessionId(), release_id: releaseId };
      let response;
      try {
        // Not under the turn's abort signal: an Esc after the body went out would report a filed request as unreachable.
        response = await fetch(`${serviceUrl}/reef/train`, {
          method: "POST",
          headers: { ...reefHeaders(), "content-type": "application/json" },
          body: JSON.stringify(body),
        });
      } catch (error) {
        ctx.ui.notify(`reef unreachable at ${serviceUrl}: ${message(error)}`, "error");
        return;
      }
      if (!response.ok) {
        ctx.ui.notify(`reef refused the request (HTTP ${response.status}): ${await response.text()}`, "error");
        return;
      }
      const answer = await response.json();
      ctx.ui.notify(`Training request ${answer.agent_record_id} accepted.`, "info");
    },
  });

  // The catalog oldest first; a step is a row's position in it, the creation row being 0, which is the commit
  // step the service keys the page by (a rejected step publishes nothing, so only its position names it).
  const releases = async () => {
    let response;
    try {
      response = await fetch(`${serviceUrl}/reef/harness/releases`, { headers: reefHeaders() });
    } catch (error) {
      throw new Error(`reef unreachable at ${serviceUrl}: ${message(error)}`);
    }
    if (!response.ok) throw new Error(`reef refused the catalog read (HTTP ${response.status}): ${await response.text()}`);
    const rows = (await response.json()).releases;
    return Array.isArray(rows) ? rows : [];
  };

  // A promoted row stays pending in the catalog; the promote is a later row naming it, so with the rows given
  // the pending row reads "promoted at step N".
  const verdictOf = (row, rows = []) => {
    if (row.pending) {
      const promoted = rows.findIndex(
        (other) => other.operation === "promote" && other.rollback_target_release_id === row.release_id,
      );
      return promoted >= 0 ? `promoted at step ${promoted}` : "pending";
    }
    const metrics = row.metrics && typeof row.metrics === "object" ? row.metrics : {};
    if (typeof metrics.selected === "boolean") return metrics.selected ? "selected" : "rejected";
    if (metrics.skipped) return "skipped";
    return String(row.operation || "unknown");
  };

  // The served head: the newest row that is neither pending nor a rejected or skipped step, since those publish
  // nothing and carry the head's id. The catalog's own current flag sits on the newest row, whatever it is.
  const headStep = (rows) => {
    for (let index = rows.length - 1; index >= 0; index--) {
      if (!["pending", "rejected", "skipped"].includes(verdictOf(rows[index]))) return index;
    }
    return -1;
  };

  const requestText = (row) => {
    const request = row.metrics && row.metrics.training_request;
    const text = request && typeof request.text === "string" ? request.text.trim() : "";
    if (!text) return "";
    return `"${text.length > 60 ? `${text.slice(0, 57)}...` : text}"`;
  };

  const lineOf = (step, rows) =>
    [
      String(step),
      String(rows[step].release_id || "").slice(0, 8),
      verdictOf(rows[step], rows),
      step === headStep(rows) ? "current" : "",
      requestText(rows[step]),
    ]
      .filter(Boolean)
      .join("  ");

  // What the step's model calls cost in tokens, when the endpoint reported them: the proposer's and the gate's.
  const tokenText = (row) => {
    const metrics = row.metrics || {};
    const over = (side, key) =>
      Object.values(side || {}).reduce((total, agent) => total + (Number(agent && agent[key]) || 0), 0);
    const proposerIn = Number(metrics.proposer_input_tokens) || 0;
    const proposerOut = Number(metrics.proposer_output_tokens) || 0;
    const gateIn = over(metrics.candidate_agents, "input_tokens") + over(metrics.current_agents, "input_tokens");
    const gateOut = over(metrics.candidate_agents, "output_tokens") + over(metrics.current_agents, "output_tokens");
    if (!proposerIn && !proposerOut && !gateIn && !gateOut) return "";
    return `tokens: proposer ${proposerIn} in / ${proposerOut} out, gate ${gateIn} in / ${gateOut} out`;
  };

  const pageUrl = (step) => `${serviceUrl}/reef/harness/releases/${step}/page`;
  // The token stays in the environment: the printed command names it as the variable, never its value.
  const curl = () =>
    `curl -fsS -H 'x-reef-scenario: ${scenario}' ` + (process.env.REEF_TOKEN ? '-H "Authorization: Bearer $REEF_TOKEN" ' : "");

  const installLine = (releaseId) =>
    `${curl()}'${serviceUrl}/reef/harness/install?adapter=pi&release_id=${encodeURIComponent(releaseId)}'` +
    ` | bash -s -- '${destDir}'`;

  const stepLines = (step, rows) => {
    const row = rows[step];
    const head = headStep(rows);
    const verdict = verdictOf(row, rows);
    const lines = [
      `Harness step ${step}: ${row.release_id} (${verdict}${step === head ? ", current" : ""})`,
      `page: ${pageUrl(step)}`,
      `read it: ${curl()}'${pageUrl(step)}' > harness-step-${step}.html`,
    ];
    if (verdict === "pending") {
      const body = JSON.stringify({ release_id: row.release_id });
      lines.push(
        `promote: ${curl()}-X POST -H 'content-type: application/json' -d '${body}' ` +
          `'${serviceUrl}/reef/scenarios/${encodeURIComponent(scenario)}/promote'`,
        `or from here: /reef-versions ${step} promote`,
        `trial install (replaces the tree at ${destDir}): ${installLine(row.release_id)}`,
      );
      if (head >= 0) lines.push(`back to the head: ${installLine(rows[head].release_id)}`);
    }
    const tokens = tokenText(row);
    if (tokens) lines.push(tokens);
    return lines;
  };

  pi.registerCommand("reef-versions", {
    description: "List this harness's versions, or show one: /reef-versions [step] [promote]",
    handler: async (args, ctx) => {
      const words = (args || "").trim().split(/\s+/).filter(Boolean);
      const promote = words[1] === "promote";
      // Digits only before Number(): "1e1" and "0x3" are numbers to it and no step to the catalog.
      const usable = words.length === 0 || (/^\d+$/.test(words[0]) && (words.length === 1 || (promote && words.length === 2)));
      if (!usable) {
        ctx.ui.notify("Usage: /reef-versions [step] [promote]", "warning");
        return;
      }
      const step = words.length ? Number(words[0]) : null;
      let rows;
      try {
        rows = await releases();
      } catch (error) {
        ctx.ui.notify(message(error), "error");
        return;
      }
      if (step === null) {
        ctx.ui.notify(rows.length ? rows.map((_, index) => lineOf(index, rows)).join("\n") : "no release on record", "info");
        return;
      }
      const row = rows[step];
      if (!row) {
        ctx.ui.notify(`no step ${step}: the catalog holds steps 0 to ${rows.length - 1}`, "warning");
        return;
      }
      if (!promote) {
        ctx.ui.notify(stepLines(step, rows).join("\n"), "info");
        return;
      }
      const verdict = verdictOf(row, rows);
      if (verdict.startsWith("promoted")) {
        ctx.ui.notify(`step ${step} is already ${verdict}; nothing to promote`, "warning");
        return;
      }
      if (verdict !== "pending") {
        ctx.ui.notify(`step ${step} is not pending (${verdict}); nothing to promote`, "warning");
        return;
      }
      const confirmed = await ctx.ui.confirm(
        `Promote harness step ${step}?`,
        `Release ${row.release_id} then serves every session that installs the head. Read ${pageUrl(step)} first.`,
      );
      if (!confirmed) {
        ctx.ui.notify(`step ${step} not promoted`, "info");
        return;
      }
      let response;
      try {
        response = await fetch(`${serviceUrl}/reef/scenarios/${encodeURIComponent(scenario)}/promote`, {
          method: "POST",
          headers: { ...reefHeaders(), "content-type": "application/json" },
          body: JSON.stringify({ release_id: row.release_id }),
        });
      } catch (error) {
        ctx.ui.notify(`reef unreachable at ${serviceUrl}: ${message(error)}`, "error");
        return;
      }
      if (!response.ok) {
        ctx.ui.notify(`reef refused the promote (HTTP ${response.status}): ${await response.text()}`, "error");
        return;
      }
      const answer = await response.json();
      ctx.ui.notify(
        `Promoted step ${step}: the head is now ${answer.release_id}; the update notice offers it at the next session start.`,
        "info",
      );
    },
  });
}

File: reef/train/cordis_backend/requests.py
"""What a release needs from the person: the ``requires`` items a training request carries.

A request posted to ``POST /reef/train`` may name what its change needs
from the person's machine, as ``{name, kind, check}`` items; the proposer
may add items its extension needs, and the commit that answered the request
carries the merged list under ``training_request.requires``. A release's
items are per step, so the manifest, the install script, ``reef-<adapter>
setup`` and the update notice read the union over the release's chain with
:func:`required_by`, the one rule for all four. Nothing here runs a check:
the checks run on the person's machine, after the person read them.
"""

from __future__ import annotations

import re
from collections.abc import Mapping, Sequence
from typing import Any

#: What a ``requires`` item asks of the person: an OS permission to grant, a variable to set, a service to connect.
REQUIRE_KINDS = ("permission", "env", "service")
#: A release names a handful of things to set up; a longer list is a request that should be split.
MAX_REQUIRES = 8
#: An item's name is shown, checked off and matched by name, so it keeps to the entry name pattern.
_REQUIRE_NAME = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*$")
#: An env item names a variable the wrapper and the extension read from the environment, so it is an identifier.
_ENV_NAME = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")


def parse_requires(value: object, *, limit: int | None = MAX_REQUIRES) -> list[dict[str, Any]]:
    """The ``requires`` list of a request as ``{name, kind, check?}`` records; a ValueError names the first bad item.

    ``value`` is the list as the route or a proposer gave it: at most
    ``limit`` objects (``MAX_REQUIRES`` for one request; ``None`` for a
    chain union, which the cap never bounds), each with a ``name`` matching
    the entry name pattern, a ``kind`` from ``REQUIRE_KINDS`` and an
    optional non empty ``check``; for kind ``env`` the variable named (the
    check, else the name) is a shell identifier. Unknown keys are dropped;
    ``None`` means no items."""
    if value is None:
        return []
    if not isinstance(value, Sequence) or isinstance(value, (str, bytes)):
        raise ValueError("requires must be a list of {name, kind, check} objects")
    if limit is not None and len(value) > limit:
        raise ValueError(f"requires must have at most {limit} items")
    items: list[dict[str, Any]] = []
    for index, item in enumerate(value):
        if not isinstance(item, Mapping):
            raise ValueError(f"requires[{index}] must be an object with a name and a kind")
        name, kind, check = item.get("name"), item.get("kind"), item.get("check")
        if not isinstance(name, str) or not _REQUIRE_NAME.fullmatch(name):
            raise ValueError(f"requires[{index}].name must be a non-empty string matching {_REQUIRE_NAME.pattern}")
        if kind not in REQUIRE_KINDS:
            raise ValueError(f"requires[{index}].kind must be one of {REQUIRE_KINDS}")
        if check is not None and (not isinstance(check, str) or not check.strip()):
            raise ValueError(f"requires[{index}].check must be a non-empty string when present")
        if kind == "env" and not _ENV_NAME.fullmatch(name if check is None else check):
            field = "name" if check is None else "check"
            raise ValueError(
                f"requires[{index}].{field} must be a variable name matching {_ENV_NAME.pattern} for kind env"
            )
        parsed: dict[str, Any] = {"name": name, "kind": kind}
        if check is not None:
            parsed["check"] = check
        items.append(parsed)
    return items


def merge_requires(base: Sequence[Mapping[str, Any]], added: Sequence[Mapping[str, Any]]) -> list[dict[str, Any]]:
    """``base`` then every ``added`` item whose name is new: the person's items stand, a proposer only adds."""
    merged = [dict(item) for item in base]
    names = {str(item.get("name")) for item in merged}
    for item in added:
        if str(item.get("name")) not in names:
            merged.append(dict(item))
            names.add(str(item.get("name")))
    return merged


def _row_requires(metrics: Mapping[str, Any] | None) -> list[dict[str, Any]]:
    """The ``training_request.requires`` items of one release's gate metrics; empty when the step read no request."""
    request = metrics.get("training_request") if isinstance(metrics, Mapping) else None
    requires = request.get("requires") if isinstance(request, Mapping) else None
    if not isinstance(requires, Sequence) or isinstance(requires, str):
        return []
    return [dict(item) for item in requires if isinstance(item, Mapping) and isinstance(item.get("name"), str)]


def _chain(rows: Sequence[Mapping[str, Any]], release_id: str | None) -> list[Mapping[str, Any]]:
    """The content chain of ``release_id`` through the catalog ``rows``, newest first.

    ``rows`` come oldest first, as ``GET /reef/harness/releases`` lists
    them, and the row that published an id is its oldest row; a promote or
    rollback row continues at its target, any other at its parent."""
    published: dict[str, Mapping[str, Any]] = {}
    for row in rows:
        if isinstance(row, Mapping) and isinstance(row.get("release_id"), str):
            published.setdefault(row["release_id"], row)
    chain: list[Mapping[str, Any]] = []
    seen: set[str] = set()
    current = release_id
    while isinstance(current, str) and current in published and current not in seen:
        seen.add(current)
        chain.append(published[current])
        current = published[current].get("rollback_target_release_id") or published[current].get("parent_release_id")
    return chain


def required_by(rows: Sequence[Mapping[str, Any]], release_id: str | None) -> list[dict[str, Any]]:
    """What a release needs from the person: every ``training_request.requires`` item over its chain, by name.

    A request's items are per release, so a later release whose request
    named nothing still carries the extension an earlier one added. The
    walk starts at ``release_id`` and follows its chain through the
    catalog ``rows``. Names keep the order of their oldest definition and
    the newest definition of a name wins. The one rule for the manifest,
    the install script, ``reef-<adapter> setup`` and the update notice."""
    merged: dict[str, dict[str, Any]] = {}
    for row in reversed(_chain(rows, release_id)):
        for item in _row_requires(row.get("metrics")):
            merged[item["name"]] = item
    return list(merged.values())


def ancestor_requiring_nothing(rows: Sequence[Mapping[str, Any]], release_id: str | None) -> str | None:
    """The newest release before ``release_id`` in its chain that requires nothing over its own chain.

    The one a machine with nothing set up installs: every later release
    in the chain needs an item, and the local release file records no completed setup.
    The creation row always qualifies; ``None`` when ``release_id`` has no
    ancestor in ``rows``."""
    ancestors = list(reversed(_chain(rows, release_id)))[:-1]
    # Oldest first, the union stays empty until the first row that named an item.
    latest = None
    for row in ancestors:
        if _row_requires(row.get("metrics")):
            break
        latest = str(row["release_id"])
    return latest


__all__ = [
    "MAX_REQUIRES",
    "REQUIRE_KINDS",
    "ancestor_requiring_nothing",
    "merge_requires",
    "parse_requires",
    "required_by",
]
